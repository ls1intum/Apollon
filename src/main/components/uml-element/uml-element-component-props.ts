export type UMLElementComponentProps = {
  id: string;
  children?: React.ReactNode;
};
