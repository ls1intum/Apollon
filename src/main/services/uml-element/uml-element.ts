import { DeepPartial } from 'redux';
import { UMLDiagramType } from '../../packages/diagram-type';
import { UMLElementType } from '../../packages/uml-element-type';
import { UMLRelationshipType } from '../../packages/uml-relationship-type';
import * as Apollon from '../../typings';
import { assign } from '../../utils/fx/assign';
import { IBoundary } from '../../utils/geometry/boundary';
import { Point } from '../../utils/geometry/point';
import { uuid } from '../../utils/uuid';
import { ILayer } from '../layouter/layer';
import { ILayoutable } from '../layouter/layoutable';
import { UMLElementFeatures } from './uml-element-features';
import { Direction } from './uml-element-port';

/** Interface of a `UMLElement` defining the properties persisted in the internal storage */
export interface IUMLElement {
  /** Unique Identifier of the `UMLElement` */
  id: string;
  /** Visual name of the `UMLElement` */
  name: string;
  /** Distinct type to recreate the `UMLElement` */
  type: UMLElementType | UMLRelationshipType | UMLDiagramType;
  /** Optional owner of the `UMLElement` */
  owner: string | null;
  /** Position and sizing of the `UMLElement` */
  bounds: IBoundary;
  /** Highlight the element with a specified color */
  highlight?: string;
  /** Colors of the element */
  fillColor?: string;
  strokeColor?: string;
  textColor?: string;
  /** Note to show for element's assessment */
  assessmentNote?: string;
}

export const enum ResizeFrom {
  TOPLEFT = 'topLeft',
  TOPRIGHT = 'topRight',
  BOTTOMLEFT = 'bottomLeft',
  BOTTOMRIGHT = 'bottomRight',
}

export const getPortsForElement = (element: IUMLElement): { [key in Direction]: Point } => {
  return {
    [Direction.Up]: new Point(element.bounds.width / 2, 0),
    [Direction.Right]: new Point(element.bounds.width, element.bounds.height / 2),
    [Direction.Down]: new Point(element.bounds.width / 2, element.bounds.height),
    [Direction.Left]: new Point(0, element.bounds.height / 2),
    [Direction.Upright]: new Point(element.bounds.width, element.bounds.height / 4),
    [Direction.Downright]: new Point(element.bounds.width, (3 * element.bounds.height) / 4),
    [Direction.Upleft]: new Point(0, element.bounds.height / 4),
    [Direction.Downleft]: new Point(0, (3 * element.bounds.height) / 4),
    [Direction.Topright]: new Point((3 * element.bounds.width) / 4, 0),
    [Direction.Bottomright]: new Point((3 * element.bounds.width) / 4, element.bounds.height),
    [Direction.Topleft]: new Point(element.bounds.width / 4, 0),
    [Direction.Bottomleft]: new Point(element.bounds.width / 4, element.bounds.height),
  };
};

/** Class implementation of `IUMLElement` to use inheritance at runtime */
export abstract class UMLElement implements IUMLElement, ILayoutable {
  /** `UMLElement` type specific feature flags */
  static features: UMLElementFeatures = {
    connectable: true,
    droppable: false,
    hoverable: true,
    movable: true,
    resizable: true,
    selectable: true,
    updatable: true,
    alternativePortVisualization: false,
  };

  static supportedRelationships: UMLRelationshipType[] = [];

  /** Checks whether an `IUMLElement` is of type `UMLElementType` */
  static isUMLElement = (element: IUMLElement): element is IUMLElement & { type: UMLElementType } =>
    element.type in UMLElementType;

  id = uuid();
  name = '';
  abstract type: UMLElementType | UMLRelationshipType | UMLDiagramType;
  bounds = { x: 0, y: 0, width: 200, height: 100 };
  owner = null as string | null;
  highlight?: string;
  fillColor?: string;
  strokeColor?: string;
  textColor?: string;
  assessmentNote?: string;
  resizeFrom: ResizeFrom = ResizeFrom.BOTTOMRIGHT;

  constructor(values?: DeepPartial<IUMLElement>) {
    assign<IUMLElement>(this, values);
  }

  /**
   * Clones an instance of `UMLElement`
   *
   * @param override - Override existing properties.
   */
  clone<T extends UMLElement>(override?: DeepPartial<IUMLElement>): T {
    const Constructor = this.constructor as new (values?: DeepPartial<IUMLElement>) => T;
    const values: IUMLElement = { ...this, ...override, id: uuid() };

    return new Constructor(values);
  }

  /** Serializes an `UMLElement` to an `Apollon.UMLElement` */
  serialize(children?: UMLElement[]): Apollon.UMLModelElement {
    return {
      id: this.id,
      name: this.name,
      type: this.type,
      owner: this.owner,
      bounds: this.bounds,
      highlight: this.highlight,
      fillColor: this.fillColor,
      strokeColor: this.strokeColor,
      textColor: this.textColor,
      assessmentNote: this.assessmentNote,
    };
  }

  /** Deserializes an `Apollon.UMLElement` to an `UMLElement` */
  deserialize<T extends Apollon.UMLModelElement>(values: T, children?: Apollon.UMLModelElement[]) {
    this.id = values.id;
    this.name = values.name;
    this.type = values.type;
    this.owner = values.owner || null;
    this.bounds = { ...values.bounds };
    this.highlight = values.highlight;
    this.fillColor = values.fillColor;
    this.strokeColor = values.strokeColor;
    this.textColor = values.textColor;
    this.assessmentNote = values.assessmentNote;
  }

  abstract render(canvas: ILayer): ILayoutable[];
}
