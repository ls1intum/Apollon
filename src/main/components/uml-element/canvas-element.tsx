import React, { Component, ComponentClass, SVGProps } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Components } from '../../packages/components';
import { UMLElementType } from '../../packages/uml-element-type';
import { ApollonView } from '../../services/editor/editor-types';
import { UMLContainer } from '../../services/uml-container/uml-container';
import { IUMLElement } from '../../services/uml-element/uml-element';
import { UMLElementRepository } from '../../services/uml-element/uml-element-repository';
import { ModelState } from '../store/model-state';
import { withTheme, withThemeProps } from '../theme/styles';
import { UMLElementComponentProps } from './uml-element-component-props';

const STROKE = 5;

type OwnProps = { child?: ComponentClass<UMLElementComponentProps> } & UMLElementComponentProps &
  SVGProps<SVGSVGElement>;

type StateProps = {
  hovered: boolean;
  selected: boolean;
  moving: boolean;
  interactive: boolean;
  interactable: boolean;
  element: IUMLElement;
};

type DispatchProps = {};

type Props = OwnProps & StateProps & DispatchProps & withThemeProps;

const enhance = compose<ComponentClass<OwnProps>>(
  withTheme,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(
    (state, props) => ({
      hovered: state.hovered[0] === props.id,
      selected: state.selected.includes(props.id),
      moving: state.moving.includes(props.id),
      interactive: state.interactive.includes(props.id),
      interactable: state.editor.view === ApollonView.Exporting || state.editor.view === ApollonView.Highlight,
      element: state.elements[props.id],
    }),
    {},
  ),
);

class CanvasElementComponent extends Component<Props> {
  render() {
    const {
      hovered,
      selected,
      moving,
      interactive,
      interactable,
      element,
      child: ChildComponent,
      children,
      theme,
      ...props
    } = this.props;

    let elements = null;
    if (UMLContainer.isUMLContainer(element) && ChildComponent) {
      elements = element.ownedElements.map((id) => <ChildComponent key={id} id={id} />);
    }
    const ElementComponent = Components[element.type as UMLElementType];

    const highlight =
      interactable && interactive
        ? theme.interactive.normal
        : interactable && hovered
        ? theme.interactive.hovered
        : element.highlight
        ? element.highlight
        : element.fillColor
        ? element.fillColor
        : theme.color.background;

    const selectedByList = element.selectedBy || [];
    return (
      <svg
        {...props}
        {...element.bounds}
        pointerEvents={moving ? 'none' : undefined}
        fillOpacity={moving ? 0.7 : undefined}
        fill={highlight}
      >
        <ElementComponent fillColor={highlight} element={UMLElementRepository.get(element)}>
          {elements}
        </ElementComponent>
        {children}
        {!interactable && (hovered || selected) && (
          <rect
            x={-STROKE / 2}
            y={-STROKE / 2}
            width={element.bounds.width + STROKE}
            height={element.bounds.height + STROKE}
            fill="none"
            stroke="#0064ff"
            strokeOpacity="0.2"
            strokeWidth={STROKE}
            pointerEvents="none"
          />
        )}
        {selectedByList.length > 0 && (
          <g>
            {selectedByList.map((selectedBy, index) => {
              const indicatorPosition = 'translate(' + (element.bounds.width + STROKE) + ' ' + index * 32 + ')';
              return (
                <g key={selectedBy.name + '_' + selectedBy.color} id={selectedBy.name + '_' + selectedBy.color}>
                  <rect
                    x={-STROKE / 2}
                    y={-STROKE / 2}
                    width={element.bounds.width + STROKE}
                    height={element.bounds.height + STROKE}
                    fill="none"
                    stroke={selectedBy.color}
                    strokeOpacity="0.2"
                    strokeWidth={STROKE}
                    pointerEvents="none"
                  />

                  <g transform={indicatorPosition} pointerEvents="none">
                    <rect
                      fillOpacity="0.2"
                      rx="10"
                      x="-40"
                      y="-20"
                      width="85px"
                      height="30px"
                      fill={selectedBy.color}
                    />
                    <text>
                      <tspan textAnchor="middle">
                        {selectedBy.name.length < 8 ? selectedBy.name : selectedBy.name.substring(0, 6) + '..'}
                      </tspan>
                    </text>
                  </g>
                </g>
              );
            })}
          </g>
        )}
      </svg>
    );
  }
}

export const CanvasElement = enhance(CanvasElementComponent);
