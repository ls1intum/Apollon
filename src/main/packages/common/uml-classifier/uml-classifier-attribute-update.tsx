import React, { useState } from 'react';
import styled from 'styled-components';
import { Button } from '../../../components/controls/button/button';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { StylePane } from '../../../components/style-pane/style-pane';
import { IUMLElement } from '../../../services/uml-element/uml-element';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type Props = {
  id: string;
  onRefChange: (instance: Textfield<any>) => void;
  value: string;
  onChange: (id: string, values: { name?: string; fillColor?: string; textColor?: string; lineColor?: string }) => void;
  onSubmitKeyUp: () => void;
  onDelete: (id: string) => () => void;
  element: IUMLElement;
};

const UmlAttributeUpdate = ({ id, onRefChange, value, onChange, onSubmitKeyUp, onDelete, element }: Props) => {
  const [colorOpen, setColorOpen] = useState(false);

  const toggleColor = () => {
    setColorOpen(!colorOpen);
  };

  const handleNameChange = (newName: string) => {
    onChange(id, { name: newName });
  };

  const handleDelete = () => {
    onDelete(id)();
  };

  return (
    <>
      <Flex>
        <Textfield ref={onRefChange} gutter value={value} onChange={handleNameChange} onSubmitKeyUp={onSubmitKeyUp} />
        <ColorButton onClick={toggleColor} />
        <Button color="link" tabIndex={-1} onClick={handleDelete}>
          <TrashIcon />
        </Button>
      </Flex>
      <StylePane open={colorOpen} element={element} onColorChange={onChange} fillColor textColor />
    </>
  );
};

export default UmlAttributeUpdate;
