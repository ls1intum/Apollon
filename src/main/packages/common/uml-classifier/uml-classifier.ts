import { DeepPartial } from 'redux';
import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { IUMLContainer, UMLContainer } from '../../../services/uml-container/uml-container';
import { IUMLElement, UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementFeatures } from '../../../services/uml-element/uml-element-features';
import * as Apollon from '../../../typings';
import { assign } from '../../../utils/fx/assign';
import { Text } from '../../../utils/svg/text';
import { UMLElementType } from '../../uml-element-type';
import { UMLClassifierAttribute } from './uml-classifier-attribute';
import { UMLClassifierMethod } from './uml-classifier-method';

export interface IUMLClassifier extends IUMLContainer {
  italic: boolean;
  underline: boolean;
  stereotype: string | null;
  deviderPosition: number;
}

export abstract class UMLClassifier extends UMLContainer implements IUMLClassifier {
  static features: UMLElementFeatures = {
    ...UMLContainer.features,
    droppable: false,
    resizable: 'WIDTH',
  };
  static stereotypeHeaderHeight = 50;
  static nonStereotypeHeaderHeight = 40;

  italic: boolean = false;
  underline: boolean = false;
  stereotype: string | null = null;
  deviderPosition: number = 0;

  get headerHeight() {
    return this.stereotype ? UMLClassifier.stereotypeHeaderHeight : UMLClassifier.nonStereotypeHeaderHeight;
  }

  constructor(values?: DeepPartial<IUMLClassifier>) {
    super();
    assign<IUMLClassifier>(this, values);
  }

  abstract reorderChildren(children: IUMLElement[]): string[];

  serialize(children: UMLElement[] = []): Apollon.UMLClassifier {
    return {
      ...super.serialize(children),
      type: this.type as UMLElementType,
      attributes: children.filter((x) => x instanceof UMLClassifierAttribute).map((x) => x.id),
      methods: children.filter((x) => x instanceof UMLClassifierMethod).map((x) => x.id),
    };
  }

  render(layer: ILayer, children: ILayoutable[] = []): ILayoutable[] {
    const attributes = children.filter((x): x is UMLClassifierAttribute => x instanceof UMLClassifierAttribute);
    const methods = children.filter((x): x is UMLClassifierMethod => x instanceof UMLClassifierMethod);

    const radix = 10;
    this.bounds.width = [this, ...attributes, ...methods].reduce(
      (current, child, index) =>
        Math.max(
          current,
          Math.round(
            (Text.size(layer, child.name, index === 0 ? { fontWeight: 'bold' } : undefined).width + 20) / radix,
          ) * radix,
        ),
      Math.round(this.bounds.width / radix) * radix,
    );

    let y = this.headerHeight;
    for (const attribute of attributes) {
      attribute.bounds.x = 0.5;
      attribute.bounds.y = y + 0.5;
      attribute.bounds.width = this.bounds.width - 1;
      y += attribute.bounds.height;
    }
    y += 1;
    this.deviderPosition = y;
    for (const method of methods) {
      method.bounds.x = 0.5;
      method.bounds.y = y + 0.5;
      method.bounds.width = this.bounds.width - 1;
      y += method.bounds.height;
    }

    this.bounds.height = y + 1;
    return [this, ...attributes, ...methods];
  }
}
