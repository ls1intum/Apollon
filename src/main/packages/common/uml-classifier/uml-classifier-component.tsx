import React, { FunctionComponent } from 'react';
import { Text } from '../../../components/controls/text/text';
import { UMLClassifier } from './uml-classifier';
import { ThemedPath, ThemedRect } from '../../../components/theme/themedComponents';

export const UMLClassifierComponent: FunctionComponent<Props> = ({ element, scale, children, fillColor }) => {
  return (
    <g>
      <ThemedRect
        fillColor={fillColor || element.fillColor}
        strokeColor="none"
        width="100%"
        height={element.stereotype ? 50 * scale : 40 * scale}
      />
      <ThemedRect
        y={element.stereotype ? 50 * scale : 40 * scale}
        width="100%"
        height={element.bounds.height - (element.stereotype ? 50 * scale : 40 * scale)}
        strokeColor="none"
      />
      {element.stereotype ? (
        <svg height={50 * scale}>
          <Text fill={element.textColor}>
            <tspan x="50%" dy={-8 * scale} textAnchor="middle" fontSize="85%">
              {`«${element.stereotype}»`}
            </tspan>
            <tspan
              x="50%"
              dy={18 * scale}
              textAnchor="middle"
              fontStyle={element.italic ? 'italic' : undefined}
              textDecoration={element.underline ? 'underline' : undefined}
            >
              {element.name}
            </tspan>
          </Text>
        </svg>
      ) : (
        <svg height={40 * scale}>
          <Text
            fill={element.textColor}
            fontStyle={element.italic ? 'italic' : undefined}
            textDecoration={element.underline ? 'underline' : undefined}
          >
            {element.name}
          </Text>
        </svg>
      )}
      {children}
      <ThemedRect width="100%" height="100%" strokeColor={element.strokeColor} fillColor="none" pointer-events="none" />
      {element.hasAttributes && (
        <ThemedPath d={`M 0 ${element.headerHeight} H ${element.bounds.width}`} strokeColor={element.strokeColor} />
      )}
      {element.hasMethods && (
        <ThemedPath d={`M 0 ${element.deviderPosition} H ${element.bounds.width}`} strokeColor={element.strokeColor} />
      )}
    </g>
  );
};

interface Props {
  element: UMLClassifier;
  scale: number;
  fillColor?: string;
  children?: React.ReactNode;
}
