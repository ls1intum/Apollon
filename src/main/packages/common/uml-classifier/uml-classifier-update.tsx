import React, { Component, ComponentClass, createRef } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import styled from 'styled-components';
import { Button } from '../../../components/controls/button/button';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { Divider } from '../../../components/controls/divider/divider';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Switch } from '../../../components/controls/switch/switch';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { Header } from '../../../components/controls/typography/typography';
import { I18nContext } from '../../../components/i18n/i18n-context';
import { localized } from '../../../components/i18n/localized';
import { ModelState } from '../../../components/store/model-state';
import { StylePane } from '../../../components/style-pane/style-pane';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { AsyncDispatch } from '../../../utils/actions/actions';
import { notEmpty } from '../../../utils/not-empty';
import { ClassElementType } from '../../uml-class-diagram';
import { UMLClassAttribute } from '../../uml-class-diagram/uml-class-attribute/uml-class-attribute';
import { UMLClassMethod } from '../../uml-class-diagram/uml-class-method/uml-class-method';
import { UMLElementType } from '../../uml-element-type';
import { UMLElements } from '../../uml-elements';
import { UMLClassifier } from './uml-classifier';
import UmlAttributeUpdate from './uml-classifier-attribute-update';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

interface OwnProps {
  element: UMLClassifier;
}

type StateProps = {};

interface DispatchProps {
  create: typeof UMLElementRepository.create;
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
  getById: (id: string) => UMLElement | null;
}

type Props = OwnProps & StateProps & DispatchProps & I18nContext;

const enhance = compose<ComponentClass<OwnProps>>(
  localized,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
    create: UMLElementRepository.create,
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
    getById: (UMLElementRepository.getById as any) as AsyncDispatch<typeof UMLElementRepository.getById>,
  }),
);

type State = {
  fieldToFocus?: Textfield<string> | null;
  colorOpen: boolean;
};

const getInitialState = (): State => ({
  fieldToFocus: undefined,
  colorOpen: false,
});

class ClassifierUpdate extends Component<Props, State> {
  state = getInitialState();
  newMethodField = createRef<Textfield<string>>();
  newAttributeField = createRef<Textfield<string>>();

  private toggleColor = () => {
    this.setState((state) => ({
      colorOpen: !state.colorOpen,
    }));
  };

  componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<{}>, snapshot?: any) {
    if (this.state.fieldToFocus) {
      this.state.fieldToFocus.focus();
      this.setState({ fieldToFocus: undefined });
    }
  }

  render() {
    const { element, getById } = this.props;
    const children = element.ownedElements.map((id) => getById(id)).filter(notEmpty);
    const attributes = children.filter((child) => child instanceof UMLClassAttribute);
    const methods = children.filter((child) => child instanceof UMLClassMethod);
    const attributeRefs: (Textfield<string> | null)[] = [];
    const methodRefs: (Textfield<string> | null)[] = [];

    return (
      <div>
        <section>
          <Flex>
            <Textfield value={element.name} onChange={this.rename(element.id)} autoFocus />
            <ColorButton onClick={this.toggleColor} />
            <Button color="link" tabIndex={-1} onClick={this.delete(element.id)}>
              <TrashIcon />
            </Button>
          </Flex>
          <StylePane
            open={this.state.colorOpen}
            element={element}
            onColorChange={this.props.update}
            fillColor
            lineColor
            textColor
          />
          <Divider />
        </section>
        <section>
          <Switch value={element.type as keyof typeof ClassElementType} onChange={this.toggle} color="primary">
            <Switch.Item value={ClassElementType.AbstractClass}>
              {this.props.translate('packages.ClassDiagram.AbstractClass')}
            </Switch.Item>
            <Switch.Item value={ClassElementType.Interface}>
              {this.props.translate('packages.ClassDiagram.Interface')}
            </Switch.Item>
            <Switch.Item value={ClassElementType.Enumeration}>
              {this.props.translate('packages.ClassDiagram.Enumeration')}
            </Switch.Item>
          </Switch>
          <Divider />
        </section>
        <section>
          <Header>{this.props.translate('popup.attributes')}</Header>
          {attributes.map((attribute, index) => (
            <UmlAttributeUpdate
              id={attribute.id}
              key={attribute.id}
              value={attribute.name}
              onChange={this.props.update}
              onSubmitKeyUp={() =>
                index === attributes.length - 1
                  ? this.newAttributeField.current?.focus()
                  : this.setState({
                      fieldToFocus: attributeRefs[index + 1],
                    })
              }
              onDelete={this.delete}
              onRefChange={(ref) => (attributeRefs[index] = ref)}
              element={attribute}
            />
          ))}
          <Textfield
            ref={this.newAttributeField}
            outline
            value=""
            onSubmit={this.create(UMLClassAttribute)}
            onSubmitKeyUp={(key: string, value: string) => {
              // if we have a value -> navigate to next field in case we want to create a new element
              if (value) {
                this.setState({
                  fieldToFocus: this.newAttributeField.current,
                });
              } else {
                // if we submit with empty value -> focus next element (either next method field or newMethodfield)
                if (methodRefs && methodRefs.length > 0) {
                  this.setState({
                    fieldToFocus: methodRefs[0],
                  });
                } else {
                  this.setState({
                    fieldToFocus: this.newMethodField.current,
                  });
                }
              }
            }}
            onKeyDown={(event) => {
              // workaround when 'tab' key is pressed:
              // prevent default and execute blur manually without switching to next tab index
              // then set focus to newAttributeField field again (componentDidUpdate)
              if (event.key === 'Tab' && event.currentTarget.value) {
                event.preventDefault();
                event.currentTarget.blur();
                this.setState({
                  fieldToFocus: this.newAttributeField.current,
                });
              }
            }}
          />
        </section>
        <section>
          <Divider />
          <Header>{this.props.translate('popup.methods')}</Header>
          {methods.map((method, index) => (
            <UmlAttributeUpdate
              id={method.id}
              key={method.id}
              value={method.name}
              onChange={this.props.update}
              onSubmitKeyUp={() =>
                index === methods.length - 1
                  ? this.newMethodField.current?.focus()
                  : this.setState({
                      fieldToFocus: methodRefs[index + 1],
                    })
              }
              onDelete={this.delete}
              onRefChange={(ref) => (methodRefs[index] = ref)}
              element={method}
            />
          ))}
          <Textfield
            ref={this.newMethodField}
            outline
            value=""
            onSubmit={this.create(UMLClassMethod)}
            onSubmitKeyUp={() =>
              this.setState({
                fieldToFocus: this.newMethodField.current,
              })
            }
            onKeyDown={(event) => {
              // workaround when 'tab' key is pressed:
              // prevent default and execute blur manually without switching to next tab index
              // then set focus to newMethodField field again (componentDidUpdate)
              if (event.key === 'Tab' && event.currentTarget.value) {
                event.preventDefault();
                event.currentTarget.blur();
                this.setState({
                  fieldToFocus: this.newMethodField.current,
                });
              }
            }}
          />
        </section>
      </div>
    );
  }

  private create = (Clazz: typeof UMLClassAttribute | typeof UMLClassMethod) => (value: string) => {
    const { element, create } = this.props;
    const member = new Clazz();
    member.name = value;
    create(member, element.id);
  };

  private rename = (id: string) => (value: string) => {
    this.props.update(id, { name: value });
  };

  private toggle = (type: keyof typeof ClassElementType) => {
    const { element, update } = this.props;
    const newType: UMLElementType = element.type === type ? ClassElementType.Class : type;
    const instance = new UMLElements[newType]({
      id: element.id,
      name: element.name,
      type: element.type,
      owner: element.owner,
      bounds: element.bounds,
      ownedElements: element.ownedElements,
    });
    update(element.id, instance);
  };

  private delete = (id: string) => () => {
    this.props.delete(id);
  };
}

export const UMLClassifierUpdate = enhance(ClassifierUpdate);
