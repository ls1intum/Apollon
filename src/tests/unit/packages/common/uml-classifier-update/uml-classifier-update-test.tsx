import { UMLElement } from '../../../../../main/services/uml-element/uml-element';
import { UMLClass } from '../../../../../main/packages/uml-class-diagram/uml-class/uml-class';
import { getRealStore } from '../../../test-utils/test-utils';
import { wrappedRender } from '../../../test-utils/render';
import * as React from 'react';
import { UMLClassifierUpdate } from '../../../../../main/packages/common/uml-classifier/uml-classifier-update';
import { UMLClassAttribute } from '../../../../../main/packages/uml-class-diagram/uml-class-attribute/uml-class-attribute';
import { UMLClassMethod } from '../../../../../main/packages/uml-class-diagram/uml-class-method/uml-class-method';

describe('test class association popup', () => {
  let elements: UMLElement[] = [];
  let umlClass: UMLClass;

  beforeEach(() => {
    // initialize  objects
    umlClass = new UMLClass({ id: 'source-test-id' });
    const umlClassAttribute = new UMLClassAttribute({
      name: 'attribute',
      owner: umlClass.id,
    });
    const umlClassMethod = new UMLClassMethod({
      name: 'classMethod',
      owner: umlClass.id,
    });
    umlClass.ownedElements = [umlClassAttribute.id, umlClassMethod.id];
    elements.push(umlClass, umlClassAttribute, umlClassMethod);
  });

  it('render', () => {
    const store = getRealStore(undefined, elements);

    const { baseElement } = wrappedRender(<UMLClassifierUpdate element={umlClass} />, { store });
    expect(baseElement).toMatchSnapshot();
  });
});
