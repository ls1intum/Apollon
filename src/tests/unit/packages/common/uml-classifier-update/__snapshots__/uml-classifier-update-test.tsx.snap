// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test class association popup render 1`] = `
<body>
  <div>
    <div>
      <section>
        <div
          class="sc-fXSgeo hRUsyw"
        >
          <input
            class="sc-bmzYkS sc-iHGNWf fEVqYL gacFBO"
            maxlength="100"
            value=""
          />
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr lgmFnr"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-dhKdcB dMUpVV"
              height="16px"
              viewBox="0 0 448 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
              />
            </svg>
          </button>
        </div>
        <hr
          class="sc-tagGq XeZYh"
        />
      </section>
      <section>
        <div
          class="sc-hmdomO gQZejn"
        >
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr koHCVn sc-bXCLTC PDrUU"
            color="primary"
          >
            Abstract
          </button>
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr koHCVn sc-bXCLTC PDrUU"
            color="primary"
          >
            Interface
          </button>
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr koHCVn sc-bXCLTC PDrUU"
            color="primary"
          >
            Enumeration
          </button>
        </div>
        <hr
          class="sc-tagGq XeZYh"
        />
      </section>
      <section>
        <h1
          class="sc-lcIPJg chQxvR"
        >
          Attributes
        </h1>
        <div
          class="sc-esYiGF jnEhCX"
        >
          <input
            class="sc-bmzYkS sc-iHGNWf fEVqYL etmSlo"
            maxlength="100"
            value="attribute"
          />
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr lgmFnr"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-dhKdcB dMUpVV"
              height="16px"
              viewBox="0 0 448 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
              />
            </svg>
          </button>
        </div>
        <input
          class="sc-bmzYkS sc-iHGNWf fEVqYL lmyIYN"
          maxlength="100"
          value=""
        />
      </section>
      <section>
        <hr
          class="sc-tagGq XeZYh"
        />
        <h1
          class="sc-lcIPJg chQxvR"
        >
          Methods
        </h1>
        <div
          class="sc-esYiGF jnEhCX"
        >
          <input
            class="sc-bmzYkS sc-iHGNWf fEVqYL etmSlo"
            maxlength="100"
            value="classMethod"
          />
          <button
            class="sc-hzhJZQ sc-fHjqPf jXCbhr lgmFnr"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-dhKdcB dMUpVV"
              height="16px"
              viewBox="0 0 448 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
              />
            </svg>
          </button>
        </div>
        <input
          class="sc-bmzYkS sc-iHGNWf fEVqYL lmyIYN"
          maxlength="100"
          value=""
        />
      </section>
    </div>
  </div>
</body>
`;
